reconcile/terraform_resources.py

```diff
@@ -368,63 +356,79 @@ def run(
-
-    # If we are not running in dry run we don't want to run with more than one account
-    if account_names and len(account_names) > 1 and not dry_run:
-        message = "Running with multiple accounts is only supported in dry-run mode"
-        logging.error(message)
-        raise MultipleAccountNamesInDryRunException(message)
+    accounts = get_aws_accounts(dry_run, account_name, exclude_accounts)
+    account_names = {a["name"] for a in accounts}
+    tf_namespaces = get_tf_namespaces(account_names)
+    if not tf_namespaces:
+        logging.warning(
+            "No terraform namespaces found, consider disabling this integration, account names: "
+            f"{', '.join(account_names)}"
+        )
 
-    ri, oc_map, tf, resource_specs = setup(
-        dry_run,
+    tf, ts = setup(
+        accounts,
+        account_names,
+        tf_namespaces,
         print_to_file,
         thread_pool_size,
-        internal,
-        use_jump_host,
-        account_names,
-        exclude_accounts,
     )
-    publish_metrics(resource_specs, QONTRACT_INTEGRATION)
+    if defer:
+        defer(tf.cleanup)
 
-    if not dry_run and oc_map and defer:
-        defer(oc_map.cleanup)
+    publish_metrics(ts.resource_spec_inventory, QONTRACT_INTEGRATION)
 
     if print_to_file:
-        cleanup_and_exit(tf)
-    if tf is None:
-        err = True
-        cleanup_and_exit(tf, err)
+        return
+
+    runner(
+        accounts=accounts,
+        account_names=account_names,
+        tf_namespaces=tf_namespaces,
+        tf=tf,
+        ts=ts,
+        dry_run=dry_run,
+        enable_deletion=enable_deletion,
+        thread_pool_size=thread_pool_size,
+        internal=internal,
+        use_jump_host=use_jump_host,
+        light=light,
+        vault_output_path=vault_output_path,
+        defer=defer,
+    )
+
 
+def runner(
+    accounts: list[dict[str, Any]],
+    account_names: set[str],
+    tf_namespaces: list[NamespaceV1],
+    tf: Terraform,
+    ts: Terrascript,
+    dry_run: bool,
+    enable_deletion: bool = False,
+    thread_pool_size: int = 10,
+    internal: Optional[bool] = None,
+    use_jump_host: bool = True,
+    light: bool = False,
+    vault_output_path: str = "",
+    defer: Optional[Callable] = None,
+) -> None:
     if not light:
         disabled_deletions_detected, err = tf.plan(enable_deletion)
         if err:
-            cleanup_and_exit(tf, err)
+            raise RuntimeError("Terraform plan has errors")
         if disabled_deletions_detected:
-            cleanup_and_exit(tf, disabled_deletions_detected)
+            raise RuntimeError("Terraform plan has disabled deletions detected")
 
     if dry_run:
-        cleanup_and_exit(tf)
+        return
 
+    acc_name = accounts[0]["name"] if accounts else None
     if not light and tf.should_apply:
         err = tf.apply()
         if err:
-            cleanup_and_exit(tf, err)
+            raise RuntimeError("Terraform apply has errors")
```

Comment:
Do you really want to display a python stacktrace here? A `logging.error` would be sufficient, or what do you think?
The same applies to the other `RuntimeErrors` above and below.

Reply:
previously it's using `sys.exit(1)` instead of exceptions, even `logging.error` here, I still need to let it exit in error state, and I remember `sys.exit(1)` is no longer recommended in this code base, right?

Reply:
I think the usage of `sys.exit(1)` is valid here.

Reply:
For current usage `sys.exit(1)` is ok, but this logic is not in root level functions, and `sys.exit(1)` means to raise `SystemExit`, which can't be caught by `expect Exception` (need `expect SystemExit`), may surprise people if any other error handing logic added later.

Any concerns regarding print additional stacktrace here? This error shouldn't happen often, and it's just more lines in log, right?

Reply:
Ok 👍